x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useState, useEffect, useRef } from 'react';
+import { isEqual } from 'lodash';
+import {
+  Map,
+  MapboxOptions,
+  NavigationControl,
+  Popup,
+  PopupOptions,
+  GeoJSONSourceOptions
+} from 'mapbox-gl';
+
+interface Props {
+  style?: React.CSSProperties;
+  getColorStyle: (scale: number) => string;
+  mapboxStyle: MapboxOptions['style'];
+  geojsonSource: NonNullable<GeoJSONSourceOptions['data']>;
+  geojsonKeyProperty: string;
+  data: Array<{ key: string; value: number; [property: string]: any }>;
+  maxValue: number;
+  renderTooltip: (props: {
+    geojsonProperties: NonNullable<mapboxgl.MapboxGeoJSONFeature['properties']>;
+    data?: Props['data'][0];
+  }) => React.ReactElement | null;
+  initialMapboxOptions?: Partial<MapboxOptions>;
+}
+export const ChoroplethMap: React.SFC<Props> = props => {
+  const {
+    style,
+    getColorStyle,
+    mapboxStyle = 'https://tiles.maps.elastic.co/styles/osm-bright/style.json',
+    geojsonSource,
+    geojsonKeyProperty,
+    data,
+    maxValue,
+    renderTooltip: ToolTip
+  } = props;
+
+  const [hoveredData, setHoveredData] = useState<Props['data'][0] | undefined>(
+    undefined
+  );
+  const [hoveredFeatureProperties, setHoveredFeatureProperties] = useState<
+    mapboxgl.MapboxGeoJSONFeature['properties']
+  >(null);
+  const hoveredFeaturePropertiesRef = useRef<
+    mapboxgl.MapboxGeoJSONFeature['properties']
+  >(null);
+  const isMapReady = useRef<boolean>(false);
+  const containerRef = useRef<HTMLDivElement>(null);
+  const toolTipRef = useRef<HTMLDivElement>(null);
+  const mapboxMap = useRef<Map | null>(null);
+  const mapboxPopup = useRef<Popup | null>(null);
+  const mapboxStyleRef = useRef(mapboxStyle);
+  const choroplethIncrementingId = useRef(0);
+  const choroplethLayerId = useRef('choropleth_regions_0');
+  const propsRef = useRef(props);
+  propsRef.current = props;
+
+  const updateChoroplethLayer = useRef(() => {
+    if (mapboxMap.current && isMapReady.current) {
+      const map = mapboxMap.current;
+      const currentLayer: mapboxgl.Layer | undefined = map.getLayer(
+        choroplethLayerId.current
+      );
+      if (currentLayer) {
+        map.removeLayer(choroplethLayerId.current);
+      }
+      const symbolLayer = (map.getStyle().layers || []).find(
+        ({ type }) => type === 'symbol'
+      );
+      choroplethIncrementingId.current++;
```

Comment:
can we store the layer id in a const and avoid storing the layer id in a ref? or do we need to increment it?